File: js/dungeon/web/start_screen.dart
part of dungeon;

class StartScreen{
  bool isOptions = false, linkClicked = false;
  final String ret = 'PRESS ESC TO RETURN',
               ent = 'NEW GAME';
  int OPEN_GAME = 1, OPEN_START = 2;
  Rect mouse = new Rect(-1, -1, 8, 8), newGame;
  String linkColor = 'rgb(213, 213, 213)';
  
  var mouseMoveE, mouseClickE;
  
  StartScreen(){
    window.requestAnimationFrame(startScreenLoop);
  }
  
  /**** Common ****/
  void drawBg([String color='#000']){
    ctx.fillStyle = color;
    ctx.fillRect(0, 0, FULLW, FULLH);
  }
  
  void drawTitle(){
    final String title = 'DUNGEON';
    
    ctx.font = "43px 'Press Start 2P'";
    final num tmpW = ctx.measureText(title).width;
    final num titleW = HALFW - tmpW/2;
    
    ctx.fillStyle = 'rgb(21, 21, 21)';                // shadow
    ctx.fillText(title, titleW + 3, 80 + 4);
    
    ctx.fillStyle = 'rgb(192, 0, 0)';                 // actual text
    ctx.fillText(title, titleW, 80);
  }
  
  
  /***** Start Screen *****/
  void startScreenLoop(num time){
    game.displayFps(time, game.oldTime);
    
    if(startScreen())
      window.requestAnimationFrame(homeScreenLoop);
    else    
      window.requestAnimationFrame(startScreenLoop);
  }
  
  bool startScreen(){  
    if(key.isPressed(KeyCode.ENTER) || DEBUG_OVERWORLD || DEBUG_LVL){
      key.lastKeyDown = KeyCode.BACKSPACE;
      mouseEvents();
      return true;
    }
    
    drawStartScreen();
    return false;
  }
  
  void drawStartScreen(){
    drawBg();
    ctx.fillRect(HALFW - 108, HALFH - 25, 228, 40); // helps with text fade
   
    util.blinkText(19, HALFW, HALFH - 5, 'ENTER COIN');
    util.blinkText(8, HALFW, HALFH + 15, '(OR PRESS ENTER)');
    
    final String strControls = 'W,A,S,D OR ARROW MOVEMENT',
                 cRite = '© 2013 JON WIEDMANN';
    
    // controls
    ctx.font = "7px 'Press Start 2P'";
    ctx.fillStyle = 'rgb(213, 213, 213)';
    ctx.fillText(strControls, HALFW - ctx.measureText(strControls).width/2, FULLH - 62);
    
    // copyright
    ctx.font = "13px 'Press Start 2P'";
    ctx.fillStyle = 'rgb(233, 233, 233)';
    ctx.fillText(cRite, HALFW - ctx.measureText(cRite).width/2, FULLH - 24);
    
    drawTitle();
  }
  
  /***** Home Screen *****/
  void homeScreenLoop(num time){
    game.displayFps(time, game.oldTime);
    
    if(homeScreen() == OPEN_GAME)
      window.requestAnimationFrame(game.gameLoop);
    else if(homeScreen() == OPEN_START)
      window.requestAnimationFrame(startScreenLoop);
    else
     window.requestAnimationFrame(homeScreenLoop);
    
  }
  
  
  int homeScreen(){
    
    if(isOptions){
      if(key.lastKeyDown == KeyCode.ESC){
        isOptions = false;
        key.lastKeyDown = KeyCode.BACKSPACE;
        mouseEvents();
      }
      
      drawOptions();
    }
    else if(linkClicked || DEBUG_OVERWORLD || DEBUG_LVL){
      key.lastDirDown = KeyCode.BACKSPACE; // prevent navigating overworld
      resetMouseEvents();
      return OPEN_GAME;
    }
    else if(key.lastKeyDown == KeyCode.O){
      isOptions = true;
      resetMouseEvents();
    }
    else if(key.lastKeyDown == KeyCode.ESC){
      resetMouseEvents();
      return OPEN_START;
    }
    else
      drawHomeScreen();
    
    
    return 0;
  }
  
  void drawHomeScreen(){
    drawBg(Color.BROWN.name);
    drawTitle();
    
    final String load = 'LOAD GAME',
                 opt = "PRESS 'o' FOR OPTIONS";
                 
    // for testing link
    //ctx.fillStyle = 'orange';
    //ctx.fillRect(newGame.x, newGame.y, newGame.w, newGame.h); 
                 
    ctx.font = "15px 'Press Start 2P'";
    ctx.fillStyle = linkColor;
    //ctx.fillText(ent, HALFW - ctx.measureText(ent).width/2, 202);
    ctx.fillText(ent, newGame.x, newGame.y + newGame.h);
    
    ctx.font = "15px 'Press Start 2P'";
    ctx.fillStyle = 'rgb(163, 163, 163)';
    ctx.fillText(load, HALFW - ctx.measureText(load).width/2, 232);
    
    ctx.font = "13px 'Press Start 2P'";
    ctx.fillStyle = 'rgb(213, 213, 213)';
    ctx.fillText(opt, HALFW - ctx.measureText(opt).width/2, 407);
    ctx.fillText(ret, HALFW - ctx.measureText(ret).width/2, 442);
  }
  void drawOptions(){
    drawBg(Color.BROWN.name);
    drawTitle();
    
    final String opt = 'OPTIONS';
    
    // Title
    ctx.font = "19px 'Press Start 2P'";
    ctx.fillStyle = 'rgb(213, 213, 213)';
    ctx.fillText(opt, HALFW - ctx.measureText(opt).width/2, 122);
    
    ctx.font = "14px 'Press Start 2P'";
    ctx.fillText('AUDIO ....................... OFF', 40, 222);
    ctx.fillText('RESOLUTION .................. 540p', 40, 262);
    
    ctx.font = "13px 'Press Start 2P'";
    ctx.fillText(ret, HALFW - ctx.measureText(ret).width/2, 442);
  }
  
  void mouseEvents(){
    newGame = new Rect(HALFW - 57, 192, 120, 13);
    
    //print('newGame: (${newGame.x}, ${newGame.y}, ${newGame.w}, ${newGame.h})');
    
    mouseMoveE = canvas.onMouseMove.listen((e){
      mouse.x = e.layerX*game.invRatio;
      mouse.y = e.layerY*game.invRatio;
      //print('mouse: (${mouse.x}, ${mouse.y})');
      
      if(util.checkCollision(mouse, newGame)){
        canvas.style.cursor = 'pointer';
        linkColor = 'blue';
      }
      else{
        canvas.style.cursor = 'auto';
        linkColor = 'rgb(213, 213, 213)';
      }
      
    });
    
    mouseClickE = canvas.onClick.listen((e){
      if(util.checkCollision(mouse, newGame))
        linkClicked = true;
    });
  }
  
  void resetMouseEvents(){
    mouseMoveE.cancel();
    mouseClickE.cancel();
    canvas.style.cursor = 'auto';
    linkColor = 'rgb(213, 213, 213)';
  }
  
}

File: games/jonsQuest/js/startScreen.js
startScreen = function(){
	
    var _time = 0,
		copyTitle1 = "JON'S",
		copyTitle2 = "QUEST",
		copyLine = "© 2013 JON WIEDMANN"
    ;
	
	function update(){
		if(13 in keysDown){
			
		    ++game.lvl;
			
			game.sound.bgMusic.start.pause();
			game.sound.bgMusic.lvl0.loop = true;
	        
			game.sound.isOn ?
	           	game.sound.bgMusic.lvl0.play() :
	           	game.sound.bgMusic.lvl0.pause();
			
			
			game.loop();		// start game
		}
		else {
		    requestAnimFrame(startScreen.loop);
		}
	}
	
	function render(){
	    ctx.fillStyle = "#000";
		ctx.fillRect(0, 0, FULLW, FULLH + game.padHUD);
		
		//---- title
		
		// title 1
		ctx.font = "29px 'Press Start 2P'";
		var startX = HALFW - ctx.measureText(copyTitle1).width / 2 + 11,
			startY = 58;
		
		ctx.setTransform(1, 0, -0.4, 1.4, 0, 0);
			ctx.fillStyle = "#222";
			ctx.fillText('J', startX+4, startY+3);
			ctx.fillStyle = "#ff6a00";
			ctx.fillText('J', startX, startY);
		ctx.setTransform(1, 0, -0.2, 1.4, 0, 0);
			ctx.fillStyle = "#222";
			ctx.fillText('O', startX+32, startY+3);
			ctx.fillStyle = "#ff6a00";
			ctx.fillText('O', startX+28, startY);
		ctx.setTransform(1, 0, 0.05, 1.41, 0, -1);
			ctx.fillStyle = "#222";
			ctx.fillText('N', startX+58, startY+3);
			ctx.fillStyle = "#ff6a00";
			ctx.fillText('N', startX+54, startY);
		ctx.setTransform(1, 0, 0.23, 1.4, 0, 0);
			ctx.fillStyle = "#222";
			ctx.fillText("'", startX+78, startY+3);
			ctx.fillStyle = "#ff6a00";
			ctx.fillText("'", startX+74, startY);
		ctx.setTransform(1, 0, 0.42, 1.4, 0, 0);
			ctx.fillStyle = "#222";
			ctx.fillText('S', startX+95, startY+3);
			ctx.fillStyle = "#ff6a00";
			ctx.fillText('S', startX+91, startY);
	
		
		// title 2
		ctx.font = "36px 'Press Start 2P'";
		startX = HALFW - ctx.measureText(copyTitle2).width / 2 + 30;
		startY = 98;
		
		ctx.setTransform(1, 0, -0.5, 1.6, 0, 0);
			ctx.fillStyle = "#222";
			ctx.fillText('Q', startX+4, startY+3);
			ctx.fillStyle = "#ff6a00";
			ctx.fillText('Q', startX, startY);
		ctx.setTransform(1, 0, -0.25, 1.6, 0, 0);
			ctx.fillStyle = "#222";
			ctx.fillText('U', startX+26, startY+3);
			ctx.fillStyle = "#ff6a00";
			ctx.fillText('U', startX+22, startY);
		ctx.setTransform(1, 0, 0.03, 1.6, 0, 0);
			ctx.fillStyle = "#222";
			ctx.fillText('E', startX+50, startY+3);
			ctx.fillStyle = "#ff6a00";
			ctx.fillText('E', startX + 46, startY);
		ctx.setTransform(1, 0, 0.25, 1.6, 0, 0);
			ctx.fillStyle = "#222";
			ctx.fillText('S', startX+74, startY+3);
			ctx.fillStyle = "#ff6a00";
			ctx.fillText('S', startX + 70, startY);
		ctx.setTransform(1, 0, 0.5, 1.6, 0, 0);
			ctx.fillStyle = "#222";
			ctx.fillText('T', startX+90, startY+4);
			ctx.fillStyle = "#ff6a00";
			ctx.fillText('T', startX + 86, startY);
		ctx.setTransform(1,0,0,1,0,0);	// reset
		
		//---- press enter
		utils.blinkText(22, HALFW, HALFH + 81)
		
		//---- copyright
		ctx.font = "13px 'Press Start 2P'"
		ctx.fillStyle = "#ddd"
		
		ctx.fillText(copyLine, HALFW - ctx.measureText(copyLine).width / 2, FULLH + 44)		
	}
	
	return {
		loop: function(){
			update()
			render()
			
			var now = new Date().getTime()    
            game.dt = now - (_time || now)		// using game.dt avoids modifying utils method
	    	_time = now
			
		}
	}
}()
